src/hotspot/share/classfile/vmClasses.cpp

```diff
@@ -0,0 +1,261 @@
+  }
+
+  assert(vmClasses::Object_klass() != NULL, "well-known classes should now be initialized");
+
+  java_lang_Object::register_natives(CHECK);
+
+  // Calculate offsets for String and Class classes since they are loaded and
+  // can be used after this point.
+  java_lang_String::compute_offsets();
+  java_lang_Class::compute_offsets();
+
+  // Fixup mirrors for classes loaded before java.lang.Class.
+  Universe::initialize_basic_type_mirrors(CHECK);
+  Universe::fixup_mirrors(CHECK);
+
+  // do a bunch more:
+  resolve_through(VM_CLASS_ID(Reference_klass), scan, CHECK);
+
+  // The offsets for jlr.Reference must be computed before
+  // InstanceRefKlass::update_nonstatic_oop_maps is called. That function uses
+  // the offsets to remove the referent and discovered fields from the oop maps,
+  // as they are treated in a special way by the GC. Removing these oops from the
+  // oop maps must be done before the usual subclasses of jlr.Reference are loaded.
+  java_lang_ref_Reference::compute_offsets();
+
+  // Preload ref klasses and set reference types
+  vmClasses::Reference_klass()->set_reference_type(REF_OTHER);
+  InstanceRefKlass::update_nonstatic_oop_maps(vmClasses::Reference_klass());
+
+  resolve_through(VM_CLASS_ID(PhantomReference_klass), scan, CHECK);
+  vmClasses::SoftReference_klass()->set_reference_type(REF_SOFT);
+  vmClasses::WeakReference_klass()->set_reference_type(REF_WEAK);
+  vmClasses::FinalReference_klass()->set_reference_type(REF_FINAL);
+  vmClasses::PhantomReference_klass()->set_reference_type(REF_PHANTOM);
+
+  // JSR 292 classes
+  VMClassID jsr292_group_start = VM_CLASS_ID(MethodHandle_klass);
+  VMClassID jsr292_group_end   = VM_CLASS_ID(VolatileCallSite_klass);
+  resolve_until(jsr292_group_start, scan, CHECK);
+  resolve_through(jsr292_group_end, scan, CHECK);
+  resolve_until(VMClassID::LIMIT, scan, CHECK);
+
+  _box_klasses[T_BOOLEAN] = vmClasses::Boolean_klass();
+  _box_klasses[T_CHAR]    = vmClasses::Character_klass();
+  _box_klasses[T_FLOAT]   = vmClasses::Float_klass();
+  _box_klasses[T_DOUBLE]  = vmClasses::Double_klass();
+  _box_klasses[T_BYTE]    = vmClasses::Byte_klass();
+  _box_klasses[T_SHORT]   = vmClasses::Short_klass();
+  _box_klasses[T_INT]     = vmClasses::Integer_klass();
+  _box_klasses[T_LONG]    = vmClasses::Long_klass();
+  //_box_klasses[T_OBJECT]  = vmClasses::object_klass();
+  //_box_klasses[T_ARRAY]   = vmClasses::object_klass();
+
+#ifdef ASSERT
+  if (UseSharedSpaces) {
+    JVMTI_ONLY(assert(JvmtiExport::is_early_phase(),
+                      "All well known classes must be resolved in JVMTI early phase"));
+    for (auto id : EnumRange<VMClassID>{}) {
+      InstanceKlass* k = _klasses[as_int(id)];
+      assert(k->is_shared(), "must not be replaced by JVMTI class file load hook");
+    }
+  }
+#endif
+}
+
+#if INCLUDE_CDS
+
+void vmClasses::resolve_shared_class(InstanceKlass* klass, ClassLoaderData* loader_data, Handle domain, TRAPS) {
```

Comment:
It isn't at all obvious to me why this functionality was relocated. I don't see anything vmClasses related in here. ??

Reply:
`vmClasses::resolve_shared_class()` is a private method and is called from here:

```
bool vmClasses::resolve(VMClassID id, TRAPS) {
  InstanceKlass** klassp = &_klasses[as_int(id)];
...
    InstanceKlass* k = *klassp;
    resolve_shared_class(k, ....);  <---- HERE
}
```

Maybe the function name is not clear enough. It's meant to "resolve a vmClass from the CDS archive". It used to be the (even more confusingly named) `SystemDictionary::quick_resolve()` method.